### PR TITLE
Format core headers

### PR DESCRIPTION
**Pull request by miscco**
It seems that the pre-commit hook does not detect extensionless headers as C++ and does not format them

### CODE REVIEW

**Comment by bernhardmgruber** [comment]
Suggestion: You may want to add this comment to `.git-blame-ignore-revs`.

**Comment by jrhemstad** [comment]
I'm guessing there must be some config setting for clang-format or pre-commit where we can teach it to recognize files without extensions as C++ headers?

**Comment by miscco** [comment]
> Suggestion: You may want to add this commit to `.git-blame-ignore-revs`.

That is not possible as we do a squash merge and that may change the commit sha

**Comment by miscco** [comment]
> I'm guessing there must be some config setting for clang-format or pre-commit where we can teach it to recognize files without extensions as C++ headers?

I did not find any possibility to change the setting. AFAIK it is hardcoded into the [pre-commit hook](https://github.com/pre-commit/mirrors-clang-format/blob/79dadb849d5247ae54fa1f13f838fde95d91b2e9/.github/workflows/main.yml#L18)

File: libcudacxx/include/cuda/std/detail/libcxx/include/__functional_base [hunk 1]
```diff
@@ -35,6 +35,7 @@
 #include <cuda/std/detail/libcxx/include/exception>
 #include <cuda/std/detail/libcxx/include/new>
 
+// comes last
 #include <cuda/std/detail/libcxx/include/__pragma_push>
```
Comment (bernhardmgruber) [inline libcudacxx/include/cuda/std/detail/libcxx/include/__functional_base:39]: Suggestion: I found a couple of these, so we may want to aim for a systematic solution. Maybe, this include could be put last using [clang-format's `IncludeCategories`](https://clang.llvm.org/docs/ClangFormatStyleOptions.html#includecategories).

**Comment by bernhardmgruber** [comment]
> > Suggestion: You may want to add this commit to `.git-blame-ignore-revs`.
> 
> That is not possible as we do a squash merge and that may change the commit sha

If you are on top of the tree and rebase & merge, I assume the hash should be preserved. But instead of taking a risk, maybe just send a follow-up PR.

File: libcudacxx/include/cuda/std/detail/libcxx/include/__functional_base (hunk 1 above)
Comment (miscco) [reply libcudacxx/include/cuda/std/detail/libcxx/include/__functional_base]: You are right, fixed

File: libcudacxx/include/cuda/std/detail/libcxx/include/ccomplex [hunk 2]
```diff
@@ -19,14 +19,4 @@
...
-#if defined(_CCCL_IMPLICIT_SYSTEM_HEADER_GCC)
-#  pragma GCC system_header
-#elif defined(_CCCL_IMPLICIT_SYSTEM_HEADER_CLANG)
-#  pragma clang system_header
-#elif defined(_CCCL_IMPLICIT_SYSTEM_HEADER_MSVC)
-#  pragma system_header
-#endif // no system header
-
-// hh 080623 Created
-
```
Comment (bernhardmgruber) [inline libcudacxx/include/cuda/std/detail/libcxx/include/ccomplex:31]: Question: This does not look like the result of clang-format. Why was this removed?

**Review by bernhardmgruber — APPROVED** [review]
I have looked through the first few headers and trust you that this changeset, except for `.clang-format` and a few cosmetic edits, only consists of automatic formatting.

File: libcudacxx/include/cuda/std/detail/libcxx/include/ccomplex (hunk 2 above)
Comment (miscco) [reply libcudacxx/include/cuda/std/detail/libcxx/include/ccomplex:31]: Because this is a pure forwarding header, which does nothing else than include another one.

In that case we mostly just not use the header guard. At the same time, I am wondering whether that is good

**Comment by griwes** [comment]
> > I'm guessing there must be some config setting for clang-format or pre-commit where we can teach it to recognize files without extensions as C++ headers?
> 
> I did not find any possibility to change the setting. AFAIK it is hardcoded into the [pre-commit hook](https://github.com/pre-commit/mirrors-clang-format/blob/79dadb849d5247ae54fa1f13f838fde95d91b2e9/.github/workflows/main.yml#L18)

That's unfortunate. Is there any reasonable way to customize the command invoked by pre-commit? If not, we should consider adding a separate thing to verify that these files are also formatted in PRs, at the very least.

File: libcudacxx/include/cuda/std/detail/libcxx/include/utility [hunk 3]
```diff
@@ -118,15 +118,14 @@ struct common_type<pair<T1, T2>, pair<U1, U2>>;
...
-                                     synth-three-way-result<T2>>
-    operator<=>(const pair<T1,T2>&, const pair<T1,T2>&);                               // C++20
+template <class T1, class T2> bool operator!=(const pair<T1,T2>&, const pair<T1,T2>&); // constexpr in C++14, removed in
+C++20 template <class T1, class T2> bool operator< (const pair<T1,T2>&, const pair<T1,T2>&); // constexpr in C++14,
+removed in C++20 template <class T1, class T2> bool operator> (const pair<T1,T2>&, const pair<T1,T2>&); // constexpr in
+C++14, removed in C++20 template <class T1, class T2> bool operator>=(const pair<T1,T2>&, const pair<T1,T2>&); //
+constexpr in C++14, removed in C++20 template <class T1, class T2> bool operator<=(const pair<T1,T2>&, const
+pair<T1,T2>&); // constexpr in C++14, removed in C++20 template <class T1, class T2> constexpr
+common_comparison_type_t<synth-three-way-result<T1>, synth-three-way-result<T2>> operator<=>(const pair<T1,T2>&, const
+pair<T1,T2>&);                               // C++20
```
Comment (griwes) [inline libcudacxx/include/cuda/std/detail/libcxx/include/utility:128]: This got mangled.

File: libcudacxx/include/cuda/std/detail/libcxx/include/utility [hunk 4]
```diff
@@ -209,7 +208,8 @@ template<class... T>
 
 template<class T, class U=T>
     constexpr T exchange(T& obj, U&& new_value)
-      noexcept(is_nothrow_move_constructible<T>::value && is_nothrow_assignable<T&, U>::value); // constexpr in C++17, noexcept in C++23
+      noexcept(is_nothrow_move_constructible<T>::value && is_nothrow_assignable<T&, U>::value); // constexpr in C++17,
+noexcept in C++23
```
Comment (griwes) [inline libcudacxx/include/cuda/std/detail/libcxx/include/utility:212]: So did this.

**Review by griwes — APPROVED** [review]
Approving, but please fix the mangled synopsis comments.

File: libcudacxx/include/cuda/std/detail/libcxx/include/utility (hunk 3 above)
Comment (miscco) [reply libcudacxx/include/cuda/std/detail/libcxx/include/utility]: I dropped the whole synposis. No one reads it and it is heavily out of date all the time

File: libcudacxx/include/cuda/std/detail/libcxx/include/ccomplex (hunk 2 above)
Comment (bernhardmgruber) [reply libcudacxx/include/cuda/std/detail/libcxx/include/ccomplex:31]: Alright, thanks!

**Review by bernhardmgruber — APPROVED** [review]

**Comment by bdice** [comment]
Hi all, @jrhemstad tagged me on this while I was out of the office. I have opened a PR that updates `.pre-commit-config.yaml` to handle these extensionless files explicitly. See that PR for more information. https://github.com/NVIDIA/cccl/pull/1676
